### PR TITLE
Cherrypick callback manager improvement fix.

### PR DESCRIPTION
**Pull request by incfly**
https://github.com/istio/istio/issues/25909
Cherrypick of the original fix as in https://github.com/envoyproxy/envoy/pull/11751.

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by incfly** [comment]
@lambdai PTAL, thanks.

File: test/integration/stats_integration_test.cc [hunk 1]
```diff
@@ -287,8 +289,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
   // If you encounter a failure here, please see
   // https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#stats-memory-tests
   // for details on how to fix.
-  EXPECT_MEMORY_EQ(m_per_cluster, 43541);
-  EXPECT_MEMORY_LE(m_per_cluster, 44600);
+  EXPECT_MEMORY_EQ(m_per_cluster, 44715);
```
Comment (lambdai) [inline test/integration/stats_integration_test.cc]: 43765
As per
```
test/integration/stats_integration_test.cc:292: Failure
Expected equality of these values:
  m_per_cluster
    Which is: 43765
  44715
```

File: test/integration/stats_integration_test.cc [hunk 2]
```diff
@@ -348,8 +352,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithRealSymbolTable) {
   // If you encounter a failure here, please see
   // https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#stats-memory-tests
   // for details on how to fix.
-  EXPECT_MEMORY_EQ(m_per_cluster, 35749);
-  EXPECT_MEMORY_LE(m_per_cluster, 36800);
+  EXPECT_MEMORY_EQ(m_per_cluster, 36827);
```
Comment (lambdai) [inline test/integration/stats_integration_test.cc]: 35973 
As per
```
[ RUN      ] IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithRealSymbolTable/IPv4
test/integration/stats_integration_test.cc:355: Failure
Expected equality of these values:
  m_per_cluster
    Which is: 35973
  36827
Stack trace:
```

**Review by lambdai — CHANGES_REQUESTED** [review]
@incfly Thank you!
Since this is cherry-pick, istio/envoy actually missed a few changes which increase the memory usage. The actually usage is fewer than in the original PR. Could you update the PR?

File: test/integration/stats_integration_test.cc (hunk 2 above)
Comment (incfly) [reply test/integration/stats_integration_test.cc]: Done.

File: test/integration/stats_integration_test.cc (hunk 1 above)
Comment (incfly) [reply test/integration/stats_integration_test.cc]: done.

**Review by lambdai — APPROVED** [review]
Awesome!

**Comment by incfly** [comment]
fails at hot restart at this line.

https://github.com/istio/envoy/blob/3d162074fca24448334e18fb2fa507ffd7819e7e/test/integration/hotrestart_test.sh#L134

```
  check [ "${CLI_HOT_RESTART_VERSION}" = "${EXPECTED_CLI_HOT_RESTART_VERSION}" ]
```

log, https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_envoy/263/test-tsan_envoy_release-1.6/43

```
TEST: Checking for match of --hot-restart-version and admin /hot_restart_version
fetching hot restart version from http://127.0.0.1:38655/hot_restart_version ...
Fetched ADMIN_HOT_RESTART_VERSION is 
     check [  = 11.104 ] ...
```

**Pull request merged**